sympy/core/tests/test_evalf.py

```diff
@@ -579,3 +580,102 @@ def test_issue_13425():
 
 def test_issue_17421():
     assert N(acos(-I + acosh(cosh(cosh(1) + I)))) == 1.0*I
+
+def test_evalfmixin():
+    x = Symbol('x')
+    y = Float('3.141592653589793238462643383279502884197', 10)
+
+    class A(EvalfMixin):
+        is_Number = False
+
+        def func(self):
+            return A
+
+        def __init__(self, a):
+            self.val = a
+
+        def _eval_evalf_options(self, prec,  options):
+            return evalf_options(self.val, prec, options)
+
+        def _eval_evalf(self, prec):
+            return self.val._evalf(prec)
+
+    class B(EvalfMixin):
+        is_Number = False
+
+        def func(self):
+            return A
+
+        def __init__(self, a):
+            self.val = a
+
+        def _eval_evalf(self, prec):
+            return self.val._evalf(prec)
+
+    assert A(FiniteSet(pi * x)).evalf(10) == FiniteSet(y * x)
+    assert B(FiniteSet(pi * x)).evalf(10) == FiniteSet(y * x)
+
+    assert A(FiniteSet(pi * x)).evalf(subs={pi * x : 1}) == FiniteSet(Float('1'))
+    assert B(FiniteSet(pi * x)).evalf(10, subs={pi * x :1}) == FiniteSet(y * x)
+
+    assert A(sin(pi*x)).evalf(subs={pi * x:0}) == 0
+    assert B(sin(pi*x)).evalf(subs={pi * x:0}) == sin(pi*x)
+
+    a = A(pi - 3.14159265358979)
+    #_eval_evalf_options are preferred over _eval_evalf if both are given.
+    assert A(a).evalf(40) != B(a).evalf(40) #_eval_evalf is less precise
+
+def test_evalf_options():
+    x = Symbol('x')
+    class X:
+
+        def __init__(self):
+            self.val = pi * x
+
+        def _eval_evalf(self, prec):
+            return self.val._evalf(prec)
+
+        def _eval_evalf_options(self, prec,  options):
+            return evalf_options(self.val, prec, options)
+
+    class A(X, EvalfMixin, Basic):
+        def could_extract_minus_sign(self):
+            return False
+
+        def as_coefficient(self, value):
+            return
+        pass
+
+    class B(X, Expr):
+        pass
+
+    class C(Basic, X):
+        def could_extract_minus_sign(self):
+            return False
+
+        def as_coefficient(self, value):
+            return
```

Comment:
Why are these methods needed? Are they used by `EvalfMixin`?

Reply:
They were needed in case of sub class of `Basic` in case of `sin`.